rfcs/0020-telemetry.md

```diff
@@ -0,0 +1,192 @@
+```
+
+Once your application collects metrics, you must emit them with a reporter. Multiple reporter libraries exist, e.g., [statsd](https://github.com/beam-telemetry/telemetry_metrics_statsd) or [CloudWatch](https://github.com/bmuller/telemetry_metrics_cloudwatch). You can also write your own, such as this [custom reporter from dotcom](https://github.com/mbta/dotcom/blob/master/lib/cms/telemetry/reporter.ex).
+
+See More: [Introduction to Telemetry in Elixir](https://blog.miguelcoba.com/introduction-to-telemetry-in-elixir).
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+As was mentioned previously, there are two phases to discuss: getting metrics out of our applications and getting those metrics into Splunk.
+
+## Getting metrics out of our applications
+
+We want to use the [telemetry](https://hexdocs.pm/telemetry/readme.html) library, [Telemetry.Metrics](https://hexdocs.pm/telemetry_metrics/Telemetry.Metrics.html), and [telemetry_poller](https://hexdocs.pm/telemetry_poller/readme.html). The telemetry library standardizes a method for emitting events. The telemetry metrics library turns those events into metrics, and the telemetry poller gathers those metrics and emits them to a reporter at whatever intervals you define. You can see an [example implementation from dotcom](https://github.com/mbta/dotcom/blob/master/lib/cms/telemetry.ex).
+
+## Getting metrics into Splunk
+
+We recommend two options for getting metrics into Splunk, and teams should feel free to pursue the path that best fits their application and expertise.
+
+**Note**
+
+There is no reason you cannot combine the two methods.
+You could use the HEC for your Elixir applications and the UF for non-Elixir applications.
+
+### Splunk Universal Forwarder
+
+The Splunk Universal Forwarder (UF) is an agent that runs alongside an application, takes in metrics in one of many formats, and forwards those metrics to Splunk.
+The UF can be used with any application, regardless of language, which makes it the most flexible way to get metrics into Splunk.
+
+It runs in a [sidecar](https://www.oreilly.com/library/view/designing-distributed-systems/9781491983638/ch02.html) to any container that emits metrics.
+
+To use it, create a metrics index in Splunk and add the sidecar to the `container_definition_json` of your `aws-ecs-container-definition` module:
+```terraform
+container_definition_json = jsonencode([
+    module.your-container.json_map_object,
+    module.sidecar-container.json_map_object
+])
+```
+
+You use the StatsD telemetry reporter highlighted above to emit your metrics in the StatsD format, and the UF will automatically forward them.
+
+#### When you should choose this method
+
+- You maintain applications that aren't written in Elixir or want all of your applications to emit metrics in a standard format.
+
+Every language has a library that emits metrics in the StatsD format; they are straightforward to set up.
```

Comment:
question: what other languages do we use which are emitting metrics in StatsD format?

Reply:
We have a dotcom monitor that is written in Node. It was written in Node because it uses Playwright to run through critical user journeys and emits the time taken as metrics. Using Playwright allows us to reuse the same scenarios for end-to-end tests, load tests, and this monitoring. We also do health checks on our dependencies and emit those as metrics: 0 for down and 1 for up. That allows us to calculate quickly calculate up-time over long periods.

Reply:
That's the only way it outputs information?